R/reactive-cached.R

```diff
@@ -0,0 +1,369 @@
+        eventFunc(),
+        function(value) {
+          req(!ignoreNULL || !isNullEvent(value))
+          isolate(keyFunc())
+        }
+      )
+    }
+
+    # Connect the (maybe) event and key with the value
+    hybrid_chain(
+      firstStep(),
+      function(cacheKeyResult) {
+        cache <- resolve_cache_object(cache, domain)
+        key_str <- digest(list(cacheKeyResult, valueExprHash), algo = "spookyhash")
+        res <- cache$get(key_str)
+
+        # Case 1: cache hit
+        if (!is.key_missing(res)) {
+          if (res$is_promise) {
+            if (res$error) {
+              return(promise_reject(valueWithVisible(res)))
+            }
+            return(promise_resolve(valueWithVisible(res)))
+
+          } else {
+            if (res$error) {
+              stop(res$value)
+            }
+            return(valueWithVisible(res))
+          }
+        }
+
+        # Case 2: cache miss
+        #
+        # valueFunc() might return a promise, or an actual value. Normally we'd
+        # use a hybrid_chain() for this, but in this case, we need to have
+        # different behavior if it's a promise or not a promise -- the
+        # information about whether or not it's a promise needs to be stored in
+        # the cache. We need to handle both cases and record in the cache
+        # whether it's a promise or not, so that any consumer of the
+        # cachedReactive() will be given the correct kind of object (a promise
+        # vs. an actual value) in the case of a future cache hit.
+        p <- withCallingHandlers(
+          withVisible(isolate(valueFunc())),
+          error = function(e) {
+            cache$set(key_str, list(
+              is_promise = FALSE,
+              value      = e,
+              visible    = TRUE,
+              error      = TRUE
+            ))
+          }
+        )
+
+        if (is.promising(p$value)) {
+          p$value <- as.promise(p$value)
+          p$value <- p$value$
+            then(function(value) {
+              res <- withVisible(value)
+              cache$set(key_str, list(
+                is_promise = TRUE,
+                value      = res$value,
+                visible    = res$visible,
+                error      = FALSE
+              ))
+              valueWithVisible(res)
+            })$
+            catch(function(e) {
+              cache$set(key_str, list(
+                is_promise = TRUE,
+                value      = e,
+                visible    = TRUE,
+                error      = TRUE
+              ))
+              stop(e)
+            })
+          valueWithVisible(p)
+        } else {
+          # result is an ordinary value, not a promise.
+          cache$set(key_str, list(
+            is_promise = FALSE,
+            value      = p$value,
+            visible    = p$visible,
+            error      = FALSE
+          ))
+          return(valueWithVisible(p))
+        }
+      })
+    },
+    domain = domain
```

Comment:
```suggestion
    }
```

Reply:
Why remove this one?

Reply:
Felt weird to have `reactive(label = label, expr, domain = domain)`. 

Thought to collect all named args at the top.